cmd/param.go

```diff
@@ -0,0 +1,176 @@
+// Copyright 2017 The kubecfg authors
+//
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+package cmd
+
+import (
+	"fmt"
+
+	"github.com/spf13/cobra"
+
+	"github.com/ksonnet/ksonnet/pkg/kubecfg"
+)
+
+const (
+	flagParamEnv       = "env"
+	flagParamComponent = "component"
+)
+
+func init() {
+	RootCmd.AddCommand(paramCmd)
+
+	paramCmd.AddCommand(paramSetCmd)
+	paramCmd.AddCommand(paramListCmd)
+	paramCmd.AddCommand(paramDiffCmd)
+
+	paramSetCmd.PersistentFlags().String(flagParamEnv, "", "Specify environment to set parameters for")
+	paramListCmd.PersistentFlags().String(flagParamEnv, "", "Specify environment to list parameters for")
+	paramDiffCmd.PersistentFlags().String(flagParamComponent, "", "Specify the component to diff against")
+}
+
+var paramCmd = &cobra.Command{
+	Use:   "param",
+	Short: `Manage ksonnet component parameters`,
+	Long: `Parameters are the customizable fields defining ksonnet components. For
+example, replica count, component name, or deployment image.
+
+Parameters are also able to be defined separately across environments. Meaning,
+this supports features to allow a "development" environment to only run a
+single replication instance for it's components, whereas allowing a "production"
+environment to run more replication instances to meet heavier production load
+demands.
+
+Environments are ksonnet "named clusters". For more information on environments,
+run:
+  ks env --help
+`,
+	RunE: func(cmd *cobra.Command, args []string) error {
+		return fmt.Errorf("Command 'param' requires a subcommand\n\n%s", cmd.UsageString())
+	},
+}
+
+var paramSetCmd = &cobra.Command{
+	Use:   "set <component-name> <param-key> <param-value>",
+	Short: "Set component or environment parameters such as replica count or name",
+	RunE: func(cmd *cobra.Command, args []string) error {
+		flags := cmd.Flags()
+		if len(args) != 3 {
+			return fmt.Errorf("'param set' takes exactly three arguments, the name of the component, and the key and value of the parameter, respectively")
+		}
+
+		component := args[0]
+		param := args[1]
+		value := args[2]
+
+		env, err := flags.GetString(flagParamEnv)
+		if err != nil {
+			return err
+		}
+
+		c := kubecfg.NewParamSetCmd(component, env, param, value)
+
```

Comment:
- [ ] Here and in each of your command functions, once you're sure that you're done rejecting problems that would be clarified by printing the usage message, you can add the following statement:
```go
cmd.SilenceUsage = true
```
After that, if you return an error from your function, Cobra won't print the usage. It's easy to miss.

Reply:
Ah, I missed that about the root command. I just happened to have been reading a few of the individual commands, and this detail came to mind.

Reply:
It turns out I can't "un-silence" usage on subcommands :/, without explicitly calling cmd.Usage() to display. I'll open an issue to consider unsetting this at the root command